### DISCUSSION TITLE
Disable 200.html and 404.html

### DISCUSSION THREAD

**Discussion by Verequies** [nuxt/nuxt#18831] [Questions]
Is it possible to disable `200.html` and `404.html` for static site generation?
I have been through the documentation and have not found anything thus far. 

I am not needing these routes at all as I handle them elsewhere with an embedded environment.

**Comment by TheAlexLichter** [comment]
Nope, [these are hardcoded](https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/core/nitro.ts#L86). But you can use e.g. the `nitro:build:before` hook to remove the routes you don't need (Access it via `nitro.options.prerender.routes`).

**Reply by Verequies** [reply]
Configuring `200.html` and `404.html` with `nitro.options.prerender.routes` does not seem to remove those files on output? Is that what is supposed to happen? 

Looking at the referenced code, it appears that when SSR is disabled it should not generate those routes, is that correct? When testing this however it still generates them and I believe SSR adds many advantages anyhow.

I believe you should have control over what files are output, hence there should be a plain and simple configuration option to disable the generation of those routes completely.

**Reply by TheAlexLichter** [reply]
> Configuring `200.html` and `404.html` with `nitro.options.prerender.routes` does not seem to remove those files on output? Is that what is supposed to happen?

The idea would be to *remove* them from the `routes` there, as they are always added by default when using SSR.

> I believe you should have control over what files are output, hence there should be a plain and simple configuration option to disable the generation of those routes completely.

You have by using e.g. the mentioned hooks. IMO the use case to disable these routes is not that common but feel free to explain and raise an issue with it! 👍🏻

**Reply by Verequies** [reply]
So, to properly disable `200.html` and `400.html` I would have to configure `nitro.options.prerender.routes`, and then use a Nuxt build hook, to delete the files after the build? And this way I could also delete the empty nitro `server` folder, as well as the unused `nitro.json`?

Regarding, the `_payload.js`, how can I rename or move it inside my configured `js` folder which contains all the JavaScript assets? I have configured Vites rollup options to separate the JS and CSS into separate folders.

**Reply by Verequies** [reply]
I have managed to do everything within `nuxt.config.ts` as below. Definitely quite a workaround for moving the `_payload.js` and removing the unneeded files when I believe those options should be builtin to either Nuxt or nitro.

```TypeScript
// Import required libraries
import { renameSync as mv, rmSync as rm } from "fs";
import { replaceInFileSync } from "replace-in-file";

// Setup path
const pwd = process.cwd();
const dist = pwd + "/dist";

// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
	"app": {
		"buildAssetsDir": "/"
	},
	"nitro": {
		"hooks": {
			"close": () => {
				// Move payload to js directory
				mv(dist.concat("/_payload.js"), dist.concat("/js/init.js"));

				// Edit index.html to point to new payload location
				replaceInFileSync({
					"files": dist.concat("/index.html"),
					"from": /\/_payload.js/g,
					"to": "/js/init.js"
				});

				// Clean up unneeded nitro files
				rm(dist.concat("/nitro.json"), { "force": true });
				rm(dist.concat("/server"), { "recursive": true, "force": true });
			}
		},
		"output": {
			"dir": dist,
			"publicDir": dist
		},
		"prerender": {
			"ignore": [
				"/200.html",
				"/404.html"
			]
		}
	},
	"ssr": true,
	"vite": {
		"build": {
			"assetsDir": ".",
			"rollupOptions": {
				"output": {
					"assetFileNames": ({ name }) => {
						if (/\.css$/.test(name ?? '')) {
							return 'css/[name]-[hash][extname]';   
						}
						
						return 'assets/[name]-[hash][extname]';                       
					},
					"chunkFileNames": "js/[name]-[hash].js",
					"entryFileNames": "js/[name]-[hash].js"
				}
			}
		}
	}
})
```

**Reply by TheAlexLichter** [reply]
> So, to properly disable `200.html` and `400.html` I would have to configure `nitro.options.prerender.routes`

You don't need to configure the routes, you can just remove the `/200` and `/400` via hook there, or even better as you did, via ignore 👍🏻

As you also found out, using hooks to then remove/move content after generation works well too!

> 
> Regarding, the `_payload.js`, how can I rename or move it inside my configured `js` folder which contains all the JavaScript assets? I have configured Vites rollup options to separate the JS and CSS into separate folders.

You can remove the `_payload.js` fully if you like.

I'm curious why you have "all the JS content" in one single payload there 👀

**Reply by Verequies** [reply]
If I remove the `_payload.js` that would break the `index.html` would it not? As it requests for that file on load. The only way to disable it would be to disable SSR functionality.

All the JS content is not actually in a single file, but under a single folder as per the below screenshot showing the build output.
![Screenshot_20230207_192226](https://user-images.githubusercontent.com/282071/217204211-4beb48bd-6ce1-4252-8287-0c9f568f9a8b.png)

I am just trying to remove as many individual files as possible. With the embedded firmware I am writing I have to include the bytes of each file within the main server code. I have this somewhat automated, but the less files the better.

**Reply by TheAlexLichter** [reply]
> I am just trying to remove as many individual files as possible. With the embedded firmware I am writing I have to include the bytes of each file within the main server code. I have this somewhat automated, but the less files the better.

Ahh, embedded firmware, I see 👍🏻

> If I remove the _payload.js that would break the index.html would it not? As it requests for that file on load. The only way to disable it would be to disable SSR functionality.

Yes, unless you chose to do API calls instead of a static payload instead - these are the default options so to say.
But as you want afully static site, this isn't feasible.

**Reply by Verequies** [reply]
What would be the API call method? 
There will be a backend API for the protocol and such, but the UI I wanted fully static so I could integrate it with Tauri, Capacitor and of course the embedded firmware.

**Reply by TheAlexLichter** [reply]
You can disable payload extraction [via the `nuxt.config.ts`](https://nuxt.com/docs/api/configuration/nuxt-config#payloadextraction). This will mean that there is no payload anymore but also no state is statically stored. Static HTML files will still be generated.

**Reply by Verequies** [reply]
I am new to SSR/SSG, what do you mean exactly by "state is statically stored"? 
Regarding generating less files, would it be possible to just have the one index.html which includes all the JS/CSS and maybe assets in the one file? How would I go about doing that?

**Reply by TheAlexLichter** [reply]
The payload.js file contains the state of the page during build time (e.g. result of API requests).

> Regarding generating less files, would it be possible to just have the one index.html which includes all the JS/CSS and maybe assets in the one file? How would I go about doing that?

You could disable all of the SSR then and have a "classic" SPA then 🤔

**Reply by Verequies** [reply]
Ah I see, good to know.
I ended up having to disable SSR as Tauri was having issues with displaying the page.
Just used `ssr: false` in the `nuxt.config.ts`. This did not merge all the JS/CSS into the one `index.html` file though?
Is there another place to do it?

**Comment by schirrel** [comment]
@TheAlexLichter @Verequies this solution above did not worked,.
I could make it work using nitro's  "prerender:generate": 
```ts
export default defineNuxtConfig({
  ssr: false,
  nitro: {
    hooks: {
      "prerender:generate"(route) {
        const routesToSkip = ['/200.html', '/404.html']; 
        if (routesToSkip.includes(route.route) ) {
          route.skip = true;
        }
      },
    },
  },
})
```

I even wrote an article about it https://dev.to/schirrel/mastering-nuxt-hooks-disable-ssg-default-pages-200-400--10n
Would be good to have it on the nuxt docs right

@TheAlexLichter  @pi0 @danielroe @atinux @antfu  what you guys think of having this on the docs? this may help other devs. If you guys allow I will create the pr

**Reply by TheAlexLichter** [reply]
PR Welcome!

**Reply by schirrel** [reply]
Thanks @TheAlexLichter @danielroe , here it is https://github.com/nuxt/nuxt/pull/30729 🙏🏻
